File: server/scraper/app/services/verdict_eval.py
"""Golden-set evaluation for the Evaluator (`POST /api/match`, the manual
"Score a Job" page's pipeline) — now also the primary regression gate for the
automatic ingest-time scoring flow, since every discovered job depends on
Evaluator correctness in a way the old RAG architecture never exposed.

Each case is scored via the exact real-usage path: only `jobDescription` is
sent (matching the manual page), letting the Analyst extract title/company
itself. The model's 6-value verdict is mapped to the golden set's 3-band
`expected` label; `INSUFFICIENT_DATA` is tracked as a separate "unscoreable"
outcome rather than silently folded into pass/fail.

Fails loud: if any case's API call errors, the whole run raises rather than
report a partial result — a run over fewer cases than expected is a
different, incomparable denominator, not a smaller version of the same
measurement.

Run: python -m app.cli eval-verdict [--runs N]
"""
import asyncio
import json
import logging
from pathlib import Path

from app.config import Settings
from app.services import match_client

logger = logging.getLogger(__name__)

GOLDEN_SET_PATH = Path(__file__).resolve().parents[2] / "tests" / "fixtures" / "golden-set.json"

# Space calls out under the "match" rate-limit bucket (10/min) — 11 cases at
# 3s apart stays comfortably inside a minute even accounting for call latency.
PACING_SECONDS = 3.0

VERDICT_BAND: dict[str, str] = {
    "STRONG_YES": "strong",
    "YES": "strong",
    "MAYBE": "weak",
    "NO": "reject",
    "STRONG_NO": "reject",
}
UNSCOREABLE_LABEL = "Unscoreable (unrecognized verdict)"


def load_golden_set() -> list[dict]:
    data = json.loads(GOLDEN_SET_PATH.read_text(encoding="utf-8"))
    return data["cases"]


def classify(verdict: str | None) -> str:
    """Map a raw 6-value verdict to the golden set's 3-band vocabulary, or
    the sentinel "unscoreable" for INSUFFICIENT_DATA / anything unrecognized."""
    return VERDICT_BAND.get((verdict or "").strip(), "unscoreable")


async def run_eval(settings: Settings, cases: list[dict] | None = None) -> list[dict]:
    """Score every golden-set case sequentially (paced) and return one result
    dict per case. Raises RuntimeError immediately if any call fails — never
    returns a partial result."""
    cases = cases if cases is not None else load_golden_set()
    results = []
    for i, case in enumerate(cases):
        if i > 0:
            await asyncio.sleep(PACING_SECONDS)
        response = await match_client.score_job(settings, case["jdText"])
        if response is None:
            raise RuntimeError(
                f"eval-verdict: API call failed for case '{case['id']}' — "
                "aborting the whole run rather than reporting a partial result."
            )
        verdict = response.get("verdict")
        band = classify(verdict)
        results.append({
            "id": case["id"],
            "title": case.get("title"),
            "expected": case["expected"],
            "uncertain": bool(case.get("uncertain", False)),
            "tags": case.get("tags") or [],
            "verdict": verdict,
            "overallScore": response.get("overallScore"),
            "band": band,
            "passed": band == case["expected"],
        })
    return results


def format_report(results: list[dict]) -> str:
    total = len(results)
    passed = sum(1 for r in results if r["passed"])
    certain = [r for r in results if not r["uncertain"]]
    uncertain = [r for r in results if r["uncertain"]]

    lines = [
        f"Evaluator golden-set report — {passed}/{total} passed",
    ]
    if certain:
        c_passed = sum(1 for r in certain if r["passed"])
        lines.append(f"  Certain cases:   {c_passed}/{len(certain)} passed")
    if uncertain:
        u_passed = sum(1 for r in uncertain if r["passed"])
        lines.append(f"  Uncertain cases: {u_passed}/{len(uncertain)} passed (reported separately, still counted above)")
    lines.append("")

    unscoreable = [r for r in results if r["band"] == "unscoreable"]
    if unscoreable:
        lines.append(UNSCOREABLE_LABEL + ":")
        for r in unscoreable:
            lines.append(f"  {r['id']}: {r['title']} — verdict={r['verdict']}")
        lines.append("")

    failures = [r for r in results if not r["passed"]]
    if failures:
        lines.append("FAILURES, grouped by tag:")
        by_tag: dict[str, list[dict]] = {}
        for r in failures:
            for tag in (r["tags"] or ["(untagged)"]):
                by_tag.setdefault(tag, []).append(r)
        for tag in sorted(by_tag):
            lines.append(f"  [{tag}]")
            for r in by_tag[tag]:
                flag = " (uncertain)" if r["uncertain"] else ""
                lines.append(
                    f"    {r['id']}: expected {r['expected']}, got {r['band']} "
                    f"(verdict={r['verdict']}, score={r['overallScore']}){flag} — {r['title']}"
                )
        lines.append("")
    else:
        lines.append("No failures.")

    return "\n".join(lines)


def format_multi_run_report(all_results: list[list[dict]]) -> str:
    """Noise-baseline report across N repeated runs of the same case set:
    aggregate pass-rate spread, plus per-case flakiness (a case whose
    pass/fail outcome changed across runs)."""
    n_runs = len(all_results)
    total = len(all_results[0]) if all_results else 0
    pass_rates = [sum(1 for r in run if r["passed"]) for run in all_results]

    lines = [
        f"Evaluator golden-set — {n_runs} runs x {total} cases",
        f"Pass rates: {pass_rates} (spread: {max(pass_rates) - min(pass_rates)} / {total})",
        "",
    ]

    by_id: dict[str, list[dict]] = {}
    for run in all_results:
        for r in run:
            by_id.setdefault(r["id"], []).append(r)

    flaky = [
        (case_id, runs) for case_id, runs in by_id.items()
        if len({r["passed"] for r in runs}) > 1
    ]
    if flaky:
        lines.append("FLAKY cases (pass/fail outcome changed across runs):")
        for case_id, runs in flaky:
            outcomes = ["pass" if r["passed"] else "fail" for r in runs]
            lines.append(f"  {case_id}: {outcomes} — {runs[0]['title']}")
        lines.append("")
    else:
        lines.append("No flaky cases — every case's pass/fail outcome was identical across all runs.")
        lines.append("")

    lines.append("Per-run reports:")
    for i, run in enumerate(all_results, start=1):
        lines.append(f"--- run {i}/{n_runs} ---")
        lines.append(format_report(run))

    return "\n".join(lines)

File: server/api/src/EvalHarness/VerdictEval.cs
using System.Text;
using System.Text.Json;

namespace ApplicationTracker.EvalHarness;

public sealed record GoldenCase
{
    public string Id { get; init; } = "";
    public string Title { get; init; } = "";
    public string Expected { get; init; } = "";
    public string JdText { get; init; } = "";
    public bool Uncertain { get; init; }
    public List<string> Tags { get; init; } = [];
}

public sealed record VerdictResult
{
    public required GoldenCase Case { get; init; }
    public string? Verdict { get; init; }
    public int? Score { get; init; }
    public required string Band { get; init; }
    public bool Passed { get; init; }
}

/// <summary>
/// Golden-set evaluation for the Evaluator.
/// </summary>
/// <remarks>
/// The primary regression gate for the whole scoring architecture: every
/// discovered job depends on Evaluator correctness, in a way the manual page
/// alone never exposed.
///
/// The model's six-value verdict is mapped onto the golden set's three bands.
/// <c>INSUFFICIENT_DATA</c> and anything unrecognised become "unscoreable"
/// rather than being folded into a pass or a fail — a case the model could not
/// judge is a different outcome from one it judged wrongly, and averaging them
/// together hides the one that matters.
/// </remarks>
public static class VerdictEval
{
    private static readonly Dictionary<string, string> VerdictBand = new(StringComparer.Ordinal)
    {
        ["STRONG_YES"] = "strong",
        ["YES"] = "strong",
        ["MAYBE"] = "weak",
        ["NO"] = "reject",
        ["STRONG_NO"] = "reject",
    };

    // /api/match is rate-limited under the "match" bucket at 10/min, and this
    // harness is a client of it like any other.
    private static readonly TimeSpan Pacing = TimeSpan.FromSeconds(3);

    public static string Classify(string? verdict) =>
        VerdictBand.GetValueOrDefault((verdict ?? "").Trim(), "unscoreable");

    public static List<GoldenCase> Load(string fixturePath)
    {
        using var doc = JsonDocument.Parse(File.ReadAllText(fixturePath));
        var cases = new List<GoldenCase>();

        foreach (var c in doc.RootElement.GetProperty("cases").EnumerateArray())
        {
            cases.Add(new GoldenCase
            {
                Id = Str(c, "id") ?? "",
                Title = Str(c, "title") ?? "",
                Expected = Str(c, "expected") ?? "",
                JdText = Str(c, "jdText") ?? "",
                Uncertain = c.TryGetProperty("uncertain", out var u) && u.ValueKind == JsonValueKind.True,
                Tags = c.TryGetProperty("tags", out var t) && t.ValueKind == JsonValueKind.Array
                    ? [.. t.EnumerateArray().Where(x => x.ValueKind == JsonValueKind.String).Select(x => x.GetString()!)]
                    : [],
            });
        }
        return cases;
    }

    public static async Task<List<VerdictResult>> RunAsync(
        MatchClient client, List<GoldenCase> cases, CancellationToken ct = default)
    {
        var results = new List<VerdictResult>();

        for (var i = 0; i < cases.Count; i++)
        {
            var c = cases[i];
            if (i > 0) await Task.Delay(Pacing, ct);

            // Deliberately not caught. A partial run is a different denominator.
            var match = await client.ScoreAsync(c.JdText, ct);
            var band = Classify(match.Verdict);

            results.Add(new VerdictResult
            {
                Case = c,
                Verdict = match.Verdict,
                Score = match.OverallScore,
                Band = band,
                Passed = band == c.Expected,
            });

            Console.Error.WriteLine($"  [{i + 1}/{cases.Count}] {c.Id}: {match.Verdict} -> {band}");
        }

        return results;
    }

    public static string Report(List<VerdictResult> results)
    {
        var sb = new StringBuilder();
        var passed = results.Count(r => r.Passed);

        sb.AppendLine();
        sb.AppendLine($"Verdict golden set: {passed}/{results.Count} passed");

        var unscoreable = results.Where(r => r.Band == "unscoreable").ToList();
        if (unscoreable.Count > 0)
            sb.AppendLine($"  unscoreable: {unscoreable.Count} " +
                $"({string.Join(", ", unscoreable.Select(r => r.Case.Id))})");

        var failures = results.Where(r => !r.Passed).ToList();
        if (failures.Count > 0)
        {
            sb.AppendLine();
            sb.AppendLine("Failures:");
            foreach (var f in failures)
                sb.AppendLine($"    {f.Case.Id}: expected {f.Case.Expected}, got {f.Band} "
                    + $"({f.Verdict}, score {f.Score}){(f.Case.Uncertain ? "  [uncertain]" : "")}");
        }

        // Grouped by tag, because a failure concentrated in one failure-mode is
        // a different signal from the same count spread across all of them.
        var byTag = results
            .SelectMany(r => r.Case.Tags.Select(t => (Tag: t, Result: r)))
            .GroupBy(x => x.Tag)
            .OrderBy(g => g.Key);

        if (byTag.Any())
        {
            sb.AppendLine();
            sb.AppendLine("By tag:");
            foreach (var g in byTag)
                sb.AppendLine($"    {g.Key,-28} {g.Count(x => x.Result.Passed)}/{g.Count()}");
        }

        return sb.ToString();
    }

    /// <summary>
    /// Multi-run report: the pass-rate spread and which cases are flaky.
    /// </summary>
    /// <remarks>
    /// A single run cannot distinguish a real regression from model noise. The
    /// spread is what makes a later comparison meaningful — a one-case drop
    /// means nothing if the baseline already varies by two.
    /// </remarks>
    public static string MultiRunReport(List<List<VerdictResult>> runs)
    {
        var sb = new StringBuilder();
        var rates = runs.Select(r => r.Count(x => x.Passed)).ToList();

        sb.AppendLine();
        sb.AppendLine($"Verdict golden set, {runs.Count} runs of {runs[0].Count} cases");
        sb.AppendLine($"  pass counts : {string.Join(", ", rates)}");
        sb.AppendLine($"  spread      : {rates.Max() - rates.Min()} "
            + $"({(rates.Max() == rates.Min() ? "stable" : "NOISY — treat single-run diffs with care")})");

        var flaky = runs[0].Select(r => r.Case.Id)
            .Where(id => runs.Select(run => run.First(r => r.Case.Id == id).Passed).Distinct().Count() > 1)
            .ToList();

        sb.AppendLine(flaky.Count == 0
            ? "  flaky cases : none"
            : $"  flaky cases : {string.Join(", ", flaky)}");

        return sb.ToString();
    }

    private static string? Str(JsonElement e, string name) =>
        e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;
}
